src/convert.jl

```diff
@@ -0,0 +1,228 @@
+module Conversion
+
+using ..SAM
+using ..BAM
+
+using BioSequences, BioAlignments
+
+global  const op_to_code = Dict(
+    'M' => 0x00,
+    'I' => 0x01,
+    'D' => 0x02,
+    'N' => 0x03,
+    'S' => 0x04,
+    'H' => 0x05,
+    'P' => 0x06,
+    '=' => 0x07,
+    'X' => 0x08,
+    'B' => 0x09,
+    '0' => 0x0a
+)
+
+function parse_cigar(record::SAM.Record)
+    
+    cigar_ops = ('M','I','D','N','S','H','P','=','X')
+    out = NamedTuple{(:L, :op), Tuple{Int64, Char}}[]
+
+    cigar = SAM.cigar(record)
+    start = 1
+
+    for (i,c) in enumerate(cigar)
+        if c ∈ cigar_ops
+            L = parse(Int, cigar[start:i-1])
+            op = c
+            push!(out, (L = L, op = op))
+            start = i+1
+        end
+    end
+    out
+end
+
+function Base.convert(::Type{BAM.Record}, record::SAM.Record; header = nothing)
+
+    bam_record = BAM.Record()
+    set_fixed_length_fields!(bam_record, record, header)
+    set_variable_length_fields!(bam_record, record)
+
+    bam_record
+end
+
+function refname_to_refid(refname::String, header)
+    isnothing(header) && return -1
+    refname == "*" && return -1
+    
+    refs = findall(header, "SQ")
+    for (i,ref) in enumerate(refs)
+        ref["SN"] == refname && return i-1
+    end
+    error("Couldn't find reference $(refname) in header.")
```

Comment:
This function is highly inefficient, since it needs to scan every single header for each record. This implies that perhaps there needs to be some kind of API to convert records from SAM to BAM in bulk, where the header is parsed only once, and then it builds a `Dict{String, Int}` to look up the refid from the refname.
Could you add a type annotation to `header`? Just for readability.

Reply:
Yeah, maybe SAM.Header could be extended to be able do that (it seems its "responsibility" to handle refnames/ids).